lib/spack/spack/test/spec_semantics.py

```diff
@@ -1465,7 +1465,7 @@ def test_abstract_contains_semantic(lhs, rhs, expected, mock_packages):
         # Flags (flags are a map, so for convenience we initialize a full Spec)
         # Note: the semantic is that of sv variants, not mv variants
         (Spec, "cppflags=-foo", "cppflags=-bar", (False, False, False)),
-        (Spec, "cppflags='-bar -foo'", "cppflags=-bar", (False, False, False)),
+        (Spec, "cppflags='-bar -foo'", "cppflags=-bar", (False, True, False)),
```

Comment:
(Test with wrong semantic) This test seems logically wrong. If one of the two specs satisfies the other ($\subseteq$ semantic) they must intersect ($\cap$ semantic):
```suggestion
        (Spec, "cppflags='-bar -foo'", "cppflags=-bar", (True, True, False)),
```

Reply:
I have to change the implementation of intersects for this to work.

Right now (i.e. outside of this PR), it is possible for there to be satisfaction without intersection: that is because `intersects` examines `.propagate`, and `satisfies` does not.

With that in mind I want to make sure we agree on the desired semantics:

* Ignoring `.propagate` I think the definition of `intersects` for a `FlagMap` should just be `return True` (because the point of this PR is to allow any combination of flags, including `-O2` and `-O3`, for example).
* I'm not clear why propagation is considered as a part of intersection, so I still think `.intersects` should always return `True`:
  * Why should `cflags==-O2` not intersect with `cflags=-O2`? My understanding is that `intersects` should only return `False` if the specs conflict with each other. Not only do these not conflict, I think they satisfy each other 
    * Or at least, `==-O2` satisfies `=-O2` (I assume this is not controversial)
    * I'd argue the other direction too:  `=-O2` satisfies `==-O2`, because in terms of the immediate spec, the result is the same (in a reuse circumstance for example, this wouldn't matter because regardless of propagation, satisfaction would be resolved via the dependencies).
  * This was introduced in https://github.com/spack/spack/commit/bc209c470d07621ba9af277dab71d048d11f2844 and continued in https://github.com/spack/spack/commit/d54611af2c56a04ed0990055d357b30b54b30a5d (the commit responsible for the creation of `intersects`, written by you, so I think you're the person to litigate this with)

Reply:
In https://github.com/spack/spack/commit/d54611af2c56a04ed0990055d357b30b54b30a5d I just split the function trying to keep previous semantic.

Reply:
OK, `.intersects` now always returns `True`, and I have done my best to also update the accompanying `.constrain` function to achieve consistency (see new test `test_constrain_compiler_flags`).